pinot-core/src/test/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManagerTest.java

```diff
@@ -609,6 +639,43 @@ public void testFileRemovedDuringOnlineTransition()
     segmentDataManager._state.set(segmentDataManager, LLRealtimeSegmentDataManager.State.HOLDING);
     segmentDataManager.goOnlineFromConsuming(metadata);
     Assert.assertFalse(new File(segTarFileName).exists());
+    segmentDataManager.destroy();
+  }
+
+  @Test
+  public void testOnlyOneSegmentHoldingTheSemaphoreForParticularPartition()
+      throws Exception {
+    long timeout = 2_000L;
+    FakeLLRealtimeSegmentDataManager firstSegmentDataManager = createFakeSegmentManager();
+    Assert.assertTrue(firstSegmentDataManager.getAcquiredConsumerSemaphore().get());
+    Assert.assertEquals(firstSegmentDataManager.getPartitionConsumerSemaphore().availablePermits(), 0);
+
+    // Release semaphore after timeout.
+    Thread releaseSemaphore = new Thread(() -> {
+      try {
+        Thread.sleep(timeout);
+      } catch (InterruptedException e) {
+        throw new RuntimeException("InterruptedException when sleeping for " + timeout + "ms");
+      } finally {
+        firstSegmentDataManager.destroy();
+      }
+    });
+    releaseSemaphore.start();
+
+    // The second segment will be hanged until the semaphore gets released.
+    FakeLLRealtimeSegmentDataManager secondSegmentDataManager = createFakeSegmentManager();
+    Assert.assertEquals(firstSegmentDataManager.getPartitionConsumerSemaphore(),
```

Comment:
we have the global semaphore map right here in this test, so you can use that instead of adding getters that are visible for testing?